Reader: enable ESC key navigation on close following-search

psealock opened the pull request:
**Issue**
Keyboard navigation on "ESC" does not close following-search box in Reader.

<img width="518" alt="screen shot 2016-05-19 at 9 48 05 pm" src="https://cloud.githubusercontent.com/assets/1922453/15389513/6fca5dfc-1e0b-11e6-99dc-722ec4643d91.png">

**Steps to Reproduce**
* navigate to ```http://calypso.localhost:3000/following/edit```
* click the second magnifying glass to search the sites you follow
<img width="481" alt="screen shot 2016-05-19 at 9 49 34 pm" src="https://cloud.githubusercontent.com/assets/1922453/15389567/b346180a-1e0b-11e6-8ca4-d0eff4c9d1a4.png">
* enter text
* press "Escape" key

**What does happen**
* search query is cleared and all the blogs followed return below

**What should happen**
* it should additionally return to original menu

**Problem**
 ```props.pinned``` was designated to denote a search box that was always open, and included styles involving its position. It also controlled logic around the component's ```state.isOpen```. This caused multiple sources of truth.

The easy fix would have been to set ```pinned: { true }``` and override the styles and pass the buck. 

**Fix**
* Use ```props.isOpen``` as the single method of dictating the component's state from outside.
* Allow ```onSearchClose``` to always be available.

@bluefuton

blowery commented on client/components/search/index.jsx:
```diff
@@ -201,15 +198,11 @@ const Search = React.createClass( {
 
 	keyUp: function( event ) {
-		if ( event.which === 13 && isMobile() ) {
+		if ( event.key === 'Enter' && isMobile() ) {
```
keyboard events are so fun. 

should probably check for key and use it if it's there, otherwise fallback to which.

Is react doing any normalization of the event?

psealock replied:
>keyboard events are so fun.

So much! I dug around a bit and, yes React does a whole bunch of normalization. 

The **TL;DR** on below is, when using keyup/keydown, ```event.key``` is reliable for non character keys such as ```Enter``` and ```ArrowRight```, but detection of characters is best left to ```event.keyCode```.

First, ```event.which``` [falls back to ```event.keyCode```](https://github.com/facebook/react/blob/v0.14.3/src/renderers/dom/client/syntheticEvents/SyntheticKeyboardEvent.js#L66-L68) anyways, so may as well forget about it.

Next, ```event.keyCode``` is fully supported, but deprecated. Weird because there isn't a clear replacement, but its useable.

Finally, ```event.key``` is supported by all except Safari but not deprecated. React does some [normalization](https://github.com/facebook/react/blob/v0.14.3/src/renderers/dom/client/utils/getEventKey.js#L75-L85) (ie, "Esc" => "Escape") if ```event.key``` exists. Otherwise, [it maps ```event.keyCode```](https://github.com/facebook/react/blob/v0.14.3/src/renderers/dom/client/utils/getEventKey.js#L95-L99) to the string you'd expect, but only for non-character keys.

Eesh, curiosity got the best of me on this one, maybe it deserves a write-up. Anyways, I'll leave the ```event.key``` here as is.

psealock commented:
The ```pinned``` prop was being used as a way of including styles as well as denoting a search component that is always meant to be open. This caused a divergence in the use of this component where some used internal state to control open/close, while others kept it static and controlled visibility via parent CSS, some using the ```pinned``` prop, others not.

I separated ```pinned```'s styles from its logic in favor of handing all control to ```isOpen```. Now, each use case is supported and the third variable (pinned) is eliminated.

psealock commented:
**Additional Note**
Since declaring Search Component's `this.state.isOpen` as the singular source of truth, I've run into the problem where users can't close the search unless there is a query, which allows the "X" to be visible.

I fixed this by using the `onBlur` callback where appropriate so a blur event closes the search box.

psealock commented:
**Update the second**
Instead of untangling the javascript to decouple ```pinned```'s behavior from styles, I've decided to untangle the styles instead.

adding `additionalClasses="is-expanded-to-container"` to so many files works, but is less than ideal, looking into simplifying this.

psealock commented:
After some hemming and hawing on my part, I feel this solution stops the problem from propagating but doesn't fix the underlying issue. The real fix is to eliminate `pinned` altogether in favor of `isOpen`, and as I discovered that touches a lot of files and necessitates some logic changes. You can see where I was headed, if interested, by checking it out the last commit before I went revert crazy.
```
git checkout 6e33a56034f50e65959cf970fa448da85e907152
```
I propose this solution in the interim and address the `pinned`/`isOpen` power struggle in a separate PR.

On a sidenote, my `.is-expanded-to-container` class should really be something like `.is-absolutely-positioned-right`, but naming things is hard and I'm open to suggestions.

ping @bluefuton

bluefuton commented:
> I propose this solution in the interim and address the pinned/isOpen power struggle in a separate PR.

:) Sounds like a plan.

> adding additionalClasses="is-expanded-to-container" to so many files works, but is less than ideal, looking into simplifying this.

Yep, I picked up on this too. As we using it in multiple places, can we make it a boolean prop on the Search component instead?

Also worth doing a squash on the PR I think to remove the fixup commits.

psealock commented:
Thanks for the feedback @bluefuton 

**Update**
* Replaced `additionalClasses` with a Boolean prop `fitsContainer`
* Rebase => single commit

bluefuton commented:
Thanks for the updates @psealock, looks good!

One last thing: can we document the new prop in `components/search/README.md`?

psealock commented:
Thanks for the feedback. I documented the new prop in the README. 

Cheers

bluefuton commented:
Prop documentation looks great.

Is it worth making `fitsContainer` the default value? How many uses of <Search> are there in Calypso that don't need it?

psealock commented:
Ah yes, thanks. I looked through the code base and its split fairly evenly between use cases of `<Search>` that require `fitsContainer` and those that don't. There are about 12 instances of each. So I set the prop to default to `false` and added validation of the proptype. It also makes sense that `position: absolute` should be opt-in instead of opt-out.

rebased and squashed. 😄

bluefuton commented:
Nice one - it makes sense for `fitsContainer` to be `false` by default then 👍 

I noticed this while rebuilding the branch:

```WARNING: No layer found for `.search.is-expanded-to-container` of `["root", ".search.is-expanded-to-container"]` in $z-layers map. Property omitted.
	on line 218 of assets/stylesheets/shared/functions/_z-index.scss
	from line 227 of assets/stylesheets/shared/functions/_z-index.scss
	from line 104 of client/components/search/style.scss
```

Do we need to add our new class to `$z-layers`?

gwwar commented:
> Do we need to add our new class to $z-layers?

Yes, it looks like the map wasn't updated in _z-index.scss. If everything behaves normally without this being mapped, you may want to consider removing the z-index rule there. Otherwise, check out https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md#z-index and this chrome extension to help find the right z index context and value: https://github.com/gwwar/z-context

psealock commented:
@gwwar Thanks for tip and sweet devTool!

>If everything behaves normally without this being mapped, you may want to consider removing the z-index rule there.

It was a very specific rule to match the `z-index` of an open `select-dropdown` container. The only instance of these two components potentially overlapping that I found was in the Themes portion of My Sites. The behavior is normal, so I'm removing.

bluefuton commented:
Need to test in:

- [ ] section-nav component
- [ ] media library filter bar
- [ ] my sites > pages
- [ ] my sites > people
- [ ] my sites > plugins
- [ ] plugins > plugins browser
- [ ] my sites > post type filter
- [ ] my sites > posts > posts navigation
- [ ] my sites > themes
- [ ] reader > following edit
- [ ] vip > vip logs

bluefuton commented:
Looks good! 🚢

psealock commented:
@bluefuton Thanks for the thorough testing! 

Just gonna add on more as the mobile renders differently
- [x] my sites > themes (mobile view)
